src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs

```diff
@@ -3,49 +3,217 @@
+                }
+                else if (metadata == MetadataPropertyName.Values)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataMissingIdBeforeValues();
+                }
+                else
+                {
+                    Debug.Assert(metadata == MetadataPropertyName.NoMetadata);
+
+                    // Having a StartObject without metadata properties is not allowed.
+                    if (converter.ClassType == ClassType.Enumerable)
+                    {
+                        ThrowHelper.ThrowJsonException_MetadataPreservedArrayInvalidProperty(converter.TypeToConvert, reader, ref state);
+                    }
+
+                    // Skip the read of the first property name, since we already read it above.
+                    state.Current.PropertyState = StackFramePropertyState.ReadName;
+                    value = default!;
+                    return true;
+                }
+            }
+
+            if (state.Current.ObjectState == StackFrameObjectState.MetadataRefProperty)
             {
-                ThrowHelper.ThrowJsonException_MetadataValueWasNotString(reader.TokenType);
+                if (!reader.Read())
+                {
+                    value = default!;
+                    return false;
+                }
+
+                if (reader.TokenType != JsonTokenType.String)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataValueWasNotString(reader.TokenType);
+                }
+
+                string key = reader.GetString()!;
+
+                // todo: verify value is T and throw JsonException? (currently no test)
+                value = (T)state.ReferenceResolver.ResolveReferenceOnDeserialize(key);
+
+                Type propertyType = value.GetType();
+                if (propertyType.IsValueType)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataInvalidReferenceToValueType(propertyType);
+                }
+
+                state.Current.ReturnValue = value;
+                state.Current.ObjectState = StackFrameObjectState.MetadataRefPropertyEndObject;
             }
+            else if (state.Current.ObjectState == StackFrameObjectState.MetadataIdProperty)
+            {
+                if (!reader.Read())
+                {
+                    value = default!;
+                    return false;
+                }
+
+                if (reader.TokenType != JsonTokenType.String)
+                {
+                    ThrowHelper.ThrowJsonException_MetadataValueWasNotString(reader.TokenType);
+                }
 
-            MetadataPropertyName metadata = state.Current.LastSeenMetadataProperty;
-            string key = reader.GetString()!;
-            Debug.Assert(metadata == MetadataPropertyName.Id || metadata == MetadataPropertyName.Ref);
+                string id = reader.GetString()!;
+                state.Current.MetadataId = id;
```

Comment:
Yes but then the calling code gets more complicated since it still needs to call CreateObject\Collection for the non-reference cases. If we end up copying this code to the 3 callers instead as previous mentioned (for slight perf mainly to avoid some `if` statements) then it could be done at that point, although that means we add ~190LOC to the 3 locations minus some code that can be deleted (say 20) so 170*3 vs 190 LOC as it is today.